api/scripts/opensea-events.py

```diff
@@ -0,0 +1,76 @@
+from notd.store.retriever import Retriever
+from notd.store.saver import Saver
+openseaApiKey = os.environ['OPENSEA_API_KEY']
+
+
+async def test():
+    startHour = date_util.datetime_from_string('2022-07-26T23:20:00.000000')
+    endHour = date_util.datetime_from_datetime(dt=startHour, minutes=60)
+    databaseConnectionString = Database.create_psql_connection_string(username=os.environ["DB_USERNAME"], password=os.environ["DB_PASSWORD"], host=os.environ["DB_HOST"], port=os.environ["DB_PORT"], name=os.environ["DB_NAME"])
+    database = Database(connectionString=databaseConnectionString)
+    retriever = Retriever(database=database)
+    saver = Saver(database=database)
+    requester = Requester()
+    openseaRequester = Requester(headers={"Accept": "application/json", "X-API-KEY": openseaApiKey})
+    tokenListingProcessor = TokenListingProcessor(requester=requester, openseaRequester=openseaRequester)
+    registryAddress = COLLECTION_GOBLINTOWN_ADDRESS
+    queryData = {
+        'asset_contract_address': registryAddress,
+        "occurred_after": startHour,
+    }
+    await database.connect()
+    while startHour< endHour:
+        #print(startHour, 'after')
+        tokensToReprocess = []
+        queryData['occurred_after'] = startHour
+        queryData['occurred_before'] = date_util.datetime_from_datetime(dt=startHour, minutes=5)
+        response = await openseaRequester.get(url="https://api.opensea.io/api/v1/events", dataDict=queryData)
+        responseJson = response.json()
+        for asset in responseJson['asset_events']:
+            if asset['asset']:
+                # print(asset['event_timestamp'])
+                tokensToReprocess.append(asset['asset']['token_id'])
+            
+        openseaListing = await tokenListingProcessor.get_opensea_listings_for_tokens(registryAddress=registryAddress, tokenIds=tokensToReprocess)
+        async with saver.create_transaction() as connection:
+            for listing in openseaListing:
+                try:
+                    latestTokenListing = await retriever.get_token_listing_by_registry_address_token_id(registryAddress=listing.registryAddress, tokenId=listing.tokenId)
+                except NotFoundException:
+                    latestTokenListing = None
+                if not latestTokenListing:
+                    logging.info(f'Saving new listing')
+                    await saver.create_latest_token_listing(retrievedTokenListing=listing, connection=connection)
+                else:
+                    if listing.value < latestTokenListing.value:
+                        logging.info(f'Deleting existing listing')
+                        await saver.delete_latest_token_listing(latestTokenListingId=latestTokenListing.tokenListingId, connection=connection)
+                        logging.info(f'Saving listing')
+                        await saver.create_latest_token_listing(retrievedTokenListing=listing, connection=connection)
+                    
```

Comment:
this looks quite off to me. above lets say you find out that tokens #1, #2 and #3 have changed.

If #1 was cancelled and #2 and #3 were created then this block I've highlighted will only add 2 and 3 but it would not delete #1.

Reply:
(ignore the links in this comment they arent relevant, was just saying number 1, number 2...)

Reply:
Seems like we have to delete all the tokens to be reprocessed before saving the new listing